### ISSUE TITLE
Items with many bitstreams cause a proxy error when trying to view an item (anonymously) or edit an item (as administrator)

### ISSUE THREAD

**Issue by peter975** [DSpace/DSpace#8648]
**Describe the bug**
When trying to view an item with a large number of bitstreams DSpace7 will almost always generate a proxy error. DSpace7 will also report a proxy error when an admin user attempts to edit the item through the UI.

Such items will also likely be unavailable to internal (eg CRIS) or external systems that harvest content from OA repositories.

An example item with >1,000 bitstreams is here:
[https://demo7.dspace.org/items/f3b06885-d9ee-48fd-af0a-c999c1578e5c](https://demo7.dspace.org/items/f3b06885-d9ee-48fd-af0a-c999c1578e5c)
[https://demo7.dspace.org/items/f3b06885-d9ee-48fd-af0a-c999c1578e5c/edit/status](https://demo7.dspace.org/items/f3b06885-d9ee-48fd-af0a-c999c1578e5c/edit/status)

**To Reproduce**
1. Create an item in DSpace7 with a large number of bitstreams (over 1000 - the size of the individual bitstreams doesn't seem to matter, only the total number).
2. Attempt to access the item through the item page or edit the item through the item page/edit/status link.

**Expected behavior**
The item should be accessible to anonymous users without a proxy error. Anonymous users should be able to view the item and see a  list of bitstreams under the item Files section. This list of bitstreams may be paginated. 

Admin users should be able view the item in the same way as described above and should also be able to edit the item through the UI without generating a proxy error. The bitstreams should be listed (paginated) on the Edit Items --> Bitstreams tab

**Related work**
Link to any related tickets or PRs here.

**Comment by tdonohue** [comment]
@peter975 : I'm not able to reproduce this locally when I run the UI locally against the demo REST API backend (https://api7.dspace.org/server/).  However, I *am* able to reproduce this when using the demo frontend (https://demo7.dspace.org).

When I run the UI locally (on localhost), I do see that the http://localhost:4000/items/f3b06885-d9ee-48fd-af0a-c999c1578e5c page takes a REALLY long time to load.  I also see this in the server-side logs of the UI: 
```
The response for 'https://api7.dspace.org/server/api/core/items/f3b06885-d9ee-48fd-af0a-c999c1578e5c/bundles?size=9999' has the self link 'https://api7.dspace.org/server/api/core/items/f3b06885-d9ee-48fd-af0a-c999c1578e5c/bundles?embed=primaryBitstream&embed=bitstreams/format&size=1000'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=1' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=1'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=2' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=2'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=3' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=3'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=4' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=4'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=5' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=5'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=6' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=6'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=7' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=7'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=8' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=8'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=9' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=9'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=10' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=10'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=11' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=11'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=12' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=12'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=13' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=13'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=14' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=14'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=15' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=15'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=16' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=16'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=17' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=17'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=18' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=18'. These don't match. This could mean there's an issue with the REST endpoint
The response for 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?page=19' has the self link 'https://api7.dspace.org/server/api/core/bundles/6d3bf2fe-6c06-4714-96e6-f1de320cfb6d/bitstreams?embed=format&page=19'. These don't match. This could mean there's an issue with the REST endpoint
1 rules skipped due to selector errors:
  .custom-file-input:lang(en)~.custom-file-label -> unmatched pseudo-class :lang
GET /items/f3b06885-d9ee-48fd-af0a-c999c1578e5c 200 186438.635 ms - -
```

It appears (to me) that the UI is essentially paging through **all the bitstreams of the item** before loading the page (notice the final line of the logs above shows the page finally loads after 186438ms (a little over 3 minutes).  This implies that there may be some flawed code in the UI... and it's possible this "Proxy Error" is occurring when the UI's server-side processing runs out of memory (or similar).

All in all, this is **definitely a bug in the behavior of the Item page**.  It should not be paging through every bitstream of the Item. I'll pull this over to the 7.5 board and see if I can find an immediate volunteer.  

@artlowel : anyone from your team have time for this one for 7.5?

**Comment by MW3000** [comment]
I can confirm, that this is a severe issue. We have several items with a huge number of bitstreams, and I ran into the exact same issue. As you observed, @tdonohue, DSpace is not failing to finally load the page after minutes, but every reasonable proxy server preset would have timed out until then.

**Comment by amgciadev** [comment]
Hi all, that is exactly what has happened in our case - after minutes to respond for the API our Apache server times out. This is definitely a severe issue for us too as we have several items with large numbers of bitstreams - even though we encourage our depositors to rearrange the files in their submission in appropriate ways.

**Comment by tdonohue** [comment]
All, I've dug into this more today and found where the problem resides.

The issue is with our logic to determine how to set the `<meta name='citation_pdf_url'>` tag (for SEO purposes).  If you **comment out** this line of code in `metadata.service.ts` then the issue _disappears_: 
https://github.com/DSpace/dspace-angular/blob/main/src/app/core/metadata/metadata.service.ts#L154
(It might also be possible to "workaround" this bug by setting a "primary bitstream" for the Original bundle)

The issue appears to be in the `setCitationPdfUrlTag()` method which loads **every single bitstream** of the `ORIGINAL` bundle to figure out which one should be added to the `citation_pdf_url` meta tag.

@artlowel and @ybnd :  It appears this was a flaw in the logic of https://github.com/DSpace/dspace-angular/pull/1228  (it may have even existed before then).  Any chance someone on your end could find a quick fix?

FWIW, I think the quickest fix is to ONLY check the first page of bitstreams when trying to populate the `citation_pdf_url`.  If no matches are found in the first page, just add the **first listed bitstream** to `citation_pdf_url`.  By "first page", we could even just use `this.appConfig.item.bitstream.pageSize` configuration, which is simply 5 by default.  That gives sites a way to minimize the number of bitstreams checked.

**Comment by ybnd** [comment]
@tdonohue @MW3000 @peter975 looks like [this method](https://github.com/DSpace/dspace-angular/blob/d91b876bf63b6cbcaa8e0a68d524d3fac602c2ed/src/app/core/metadata/metadata.service.ts#L365-L424) is the culprit. It is called when
* there is no primary Bitstream
* the first Bitstream doesn't match any of the [allowed "citation types"](https://github.com/DSpace/dspace-angular/blob/d91b876bf63b6cbcaa8e0a68d524d3fac602c2ed/src/app/core/metadata/metadata.service.ts#L69-L76)

If the Item has no primary Bitstream and tons of Bitstreams that aren't "allowed" it would just start paging through _all of them_ like you noted above.

For a quick workaround it should be sufficient to set one of many the Bitstreams as the primary.

IMO the cleanest way to solve this would be to introduce a search method to retrieve Bitstreams by (a list of) format and use that instead of "the first one that matches".

**Comment by tdonohue** [comment]
@ybnd :  I think we should fix that method then so that it only looks at the first page.  I don't see a use case for needing to continue to page through every single bitstream looking for a match.

As noted above, I'd even be satisfied with looking only at the first _5_ bitstreams (as configured by `appConfig.item.bitstream.pageSize`), as that means we only check the bitstreams that are _displayed_ on the Item page by default.  

Basically, we should stop checking after a fixed number and just default to putting the **first bitstream in the list** in the `citation_pdf_url`.  Continually looping to try to find a PDF on the 10th or 20th page isn't really useful...as chances are that file you found isn't useful for SEO afterall.

**Comment by MW3000** [comment]
Aside from the concrete problem at hand, this highlights another issue: Why is getting 500 bitstreams via the REST API so incredibly slow. My guess is, it is somehow related to the way access permissions are retrieved? Retrieving many bitstreams was equally slow on DSpace 6.

**Comment by amgciadev** [comment]
I agree that the performance issue with the REST API itself is definitely worth addressing, especially since it seems to perform less well that older REST endpoints (e,g. 5.x endpoint). To give an example, we only noticed this issue when testing DSpace 7.4 with one of integrated systems that harvests our Apollo instance via the API. For the same set of large items, we weren't experiencing timeouts when attempting to retrieve information from them when using the 5.10 API.

**Comment by mwoodiupui** [comment]
+1 investigate the REST performance issue as well.

I've never been clear on why it is *possible* not to have a primary Bitstream.  What would be the consequences of changing the DSpace API to guarantee that one Bitstream is primary?  If the submitter doesn't pick one, then the list of "allowed citation types" could be moved to the backend and used to find one, else the first/only Bitstream is chosen.  What would this break?  What would it simplify elsewhere?  (What do we think "primary" *means*?)

**Comment by tdonohue** [comment]
I'd like to avoid this ticket spinning off into different directions.  While I'm not against investigations of REST API performance, the reason this Item page isn't loading quickly is because of the *very large number of REST requests*.  When 1,000 bitstreams exist, the DSpace UI is currently loading all 1,000 and then asking for permissions (both `canDownload` and `canRequestACopy`) on _each bitstream individually_ in order to determine whether you can download or request each bitstream.  Essentially, the primary performance issues is that for 1,000 bitstreams, the UI is making around 2,000-3,000 requests to the REST API when loading this single Item page.  Each individual request requires an SSL handshake & also some backend processing.... plus they are all being processed via SSR (server-side rendering) if you go directly to the Item page itself.

So, the fix here is to avoid making that many requests... they are not necessary.  If we just check the first page of bitstreams (e.g. 5 bitstreams shown), then we are only making around 10-15 requests (max).

Yes, I'm sure we could look at ways to also speed up each REST API request individually. But that needs to be left for a different ticket, as it's not directly related to the problem described in this ticket.  In general, the DSpace 7 REST API has been designed to make use of pagination & avoid loading large numbers of objects at once (it's not even possible to load 500 bitstreams at once in the REST API, but the UI is doing so by continually paging through every bitstream and making many many REST calls).

Therefore, the core issue is this Item page though seems to be spinning off way more requests than it needs & that causes the unexpected behaviors.

I'll keep looking for a volunteer to help us fix this.  I _think_ it should be a relatively small fix to the Item page behavior.

**Comment by artlowel** [comment]
I'm looking into it right now. The plan is to add a config option to only check the first page of bitstreams, enabled by default

**Comment by tdonohue** [comment]
@artlowel : Thanks! I'll assign this to you then.  FWIW, I'm not sure this should even be configurable.  If the bitstream you want to have in SEO isn't on the first page, you can reorder your bitstreams or (eventually) mark it as "primary" (once we have primary bitstream support working properly)

**Comment by artlowel** [comment]
Good point, if the PDF isn't on the first page you can get around it by reordering. I'll leave out the config then

**Comment by mwoodiupui** [comment]
Agreed:  REST performance is a separate issue.

**Comment by tdonohue** [comment]
Verified that https://github.com/DSpace/dspace-angular/pull/2065 fixes this main issue.  The response for Item pages with a large number of bitstreams is still slow on the demo site, but now it's more like 8 seconds instead of 3 minutes.  So, further speeding up of the REST API might be warranted, but the page should no longer crash or result in a proxy error.
